File: fs-core/src/main/java/space/sunqian/fs/utils/eventbus/EventBus.java
package space.sunqian.fs.utils.eventbus;

import space.sunqian.annotation.Nonnull;

import java.lang.reflect.Type;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.function.Consumer;

/**
 * This is the event bus interface. It provides methods to register and unregister event subscribers with the specified
 * event type, and post methods to post events with the specified event type.
 *
 * @author sunqian
 */
public interface EventBus {

    /**
     * Creates a new event bus, all event consuming will be executed in the post thread. It is same as
     * {@link #newEventBus(Executor)} with the executor: {@code Runnable::run}.
     *
     * @return a new event bus
     */
    static @Nonnull EventBus newEventBus() {
        return newEventBus(Runnable::run);
    }

    /**
     * Creates a new event bus with the specified executor.
     *
     * @param executor the executor to execute event subscribers
     * @return a new event bus
     */
    static @Nonnull EventBus newEventBus(@Nonnull Executor executor) {
        return new EventBusImpl(executor);
    }

    /**
     * Registers an event subscriber for the specified event type.
     * <p>
     * The default implementations use copy-on-write to add subscribers, so it is recommended to use
     * {@link #register(Map)} for registering multiple event subscribers at once.
     *
     * @param type       the event type
     * @param subscriber the event subscriber
     * @param <T>        the event type
     */
    <T> void register(@Nonnull Type type, @Nonnull Consumer<@Nonnull T> subscriber);

    /**
     * Registers event subscribers with the specified event types, the keys of the map are the event types, and the
     * values are the event subscribers.
     *
     * @param subscribers the event subscribers
     */
    void register(@Nonnull Map<@Nonnull Type, @Nonnull List<@Nonnull Consumer<@Nonnull ?>>> subscribers);

    /**
     * Unregisters an event subscriber which is registered via {@code register} methods. The subscriber object is the
     * same object that is passed to {@code register} methods.
     * <p>
     * The default implementations use copy-on-write to remove subscribers, so it is recommended to use
     * {@link #unregister(Iterable)} for unregistering multiple event subscribers at once.
     *
     * @param subscriber the event subscriber, which is the same object that is passed to {@code register} methods
     */
    void unregister(@Nonnull Object subscriber);

    /**
     * Unregisters event subscribers which are registered via {@code register} methods. The subscriber objects are the
     * same objects that are passed to {@code register} methods.
     *
     * @param subscribers the event subscribers, which are the same objects that are passed to {@code register} methods
     */
    void unregister(@Nonnull Iterable<?> subscribers);

    /**
     * Posts an event. This method will call {@link #post(Object, Type)} with {@code event} and
     * {@code event.getClass()}.
     *
     * @param event the event to post
     */
    default void post(@Nonnull Object event) {
        post(event, event.getClass());
    }

    /**
     * Posts an event with the specified type. This method will call {@link #post(Object, Type, DispatchMode)} with
     * {@code event}, {@code eventType}, and {@link  DispatchMode#BROADCAST}.
     *
     * @param event     the event to post
     * @param eventType the event type
     */
    default void post(@Nonnull Object event, @Nonnull Type eventType) {
        post(event, eventType, DispatchMode.BROADCAST);
    }

    /**
     * Posts an event with the specified dispatch mode. This method will call {@link #post(Object, Type, DispatchMode)}
     * with {@code event}, {@code event.getClass()}, and {@code dispatchMode}.
     *
     * @param event        the event to post
     * @param dispatchMode the dispatch mode
     */
    default void post(@Nonnull Object event, @Nonnull DispatchMode dispatchMode) {
        post(event, event.getClass(), dispatchMode);
    }

    /**
     * Posts an event with the specified type and dispatch mode.
     *
     * @param event        the event to post
     * @param eventType    the event type
     * @param dispatchMode the dispatch mode
     */
    void post(@Nonnull Object event, @Nonnull Type eventType, @Nonnull DispatchMode dispatchMode);

    /**
     * The dispatch mode for posting events.
     */
    enum DispatchMode {
        /**
         * The broadcast mode: subscribers which subscribe the event type will receive the event.
         * <p>
         * This is the default mode.
         */
        BROADCAST,
        /**
         * The chain dispatch mode: subscribers which subscribe the event type will receive the event in the order they
         * are registered, and the event propagation can be prevented by any subscriber.
         * <p>
         * In this mode, if a subscriber throws an exception, the event propagation will be stopped.
         */
        CHAIN
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/bus/EventBus.kt
package xyz.srclab.common.bus

import xyz.srclab.common.base.asTyped
import xyz.srclab.common.collect.arrayAdd
import xyz.srclab.common.collect.arrayRemove
import xyz.srclab.common.reflect.canAssignedBy
import xyz.srclab.common.run.SyncRunner
import java.util.concurrent.Executor

/**
 * Event bus. Use [register] or [registerAll] to register event handler, [unregister] and [unregisterAll] to unregister.
 *
 * [EventBusHandler] is used to handle event.
 * [EventBus] will post event for each [EventBusHandler] of which [EventBusHandler.eventType] is assignable from event,
 * in sub-to-super order.
 */
interface EventBus {

    fun register(handler: EventBusHandler<*>)

    fun registerAll(handlers: Iterable<EventBusHandler<*>>) {
        for (handler in handlers) {
            register(handler)
        }
    }

    fun unregister(handler: EventBusHandler<*>)

    fun unregisterAll(handlers: Iterable<EventBusHandler<*>>) {
        for (handler in handlers) {
            unregister(handler)
        }
    }

    fun post(event: Any)

    companion object {

        @JvmStatic
        @JvmOverloads
        fun newEventBus(executor: Executor = SyncRunner): EventBus {
            return EventBusImpl(executor)
        }

        private class EventBusImpl(
            private val executor: Executor
        ) : EventBus {

            private var handlers: Array<EventBusHandler<*>> = emptyArray()

            @Synchronized
            override fun register(handler: EventBusHandler<*>) {
                var i = 0
                while (i < handlers.size) {
                    val hi = handlers[i]
                    if (hi == handler) {
                        handlers[i] = handler
                        return
                    }
                    val ci = hi.eventType
                    val c = handler.eventType
                    if (ci == c) {
                        handlers = handlers.arrayAdd(handler, i)
                        return
                    }
                    if (ci.canAssignedBy(c)) {
                        handlers = handlers.arrayAdd(handler, i)
                        return
                    }
                    if (c.canAssignedBy(ci)) {
                        i++
                        while (i < handlers.size) {
                            val hx = handlers[i]
                            if (hx == handler) {
                                handlers[i] = handler
                                return
                            }
                            val cx = hx.eventType
                            if (cx == c) {
                                handlers = handlers.arrayAdd(handler, i)
                                return
                            }
                            if (cx.canAssignedBy(c)) {
                                handlers = handlers.arrayAdd(handler, i)
                                return
                            }
                            if (c.canAssignedBy(cx)) {
                                i++
                                continue
                            }
                            handlers = handlers.arrayAdd(handler, i)
                            return
                        }
                        handlers = handlers.arrayAdd(handler, i)
                        return
                    }
                    i++
                }
                handlers = handlers.arrayAdd(handler, 0)
            }

            @Synchronized
            override fun unregister(handler: EventBusHandler<*>) {
                var i = 0
                while (i < handlers.size) {
                    val hi = handlers[i]
                    if (hi == handler) {
                        handlers = handlers.arrayRemove(i)
                        return
                    }
                    i++
                }
            }

            override fun post(event: Any) {
                val type = event.javaClass
                for (handler in handlers) {
                    val h = handler.asTyped<EventBusHandler<Any>>()
                    val ht = handler.eventType
                    if (ht.canAssignedBy(type)) {
                        executor.execute {
                            h.doEvent(event)
                        }
                    }
                }
            }
        }
    }
}